omegaconf/_utils.py

```diff
@@ -39,6 +39,25 @@ def isint(s: str) -> bool:
 
 
 def get_yaml_loader() -> Any:
+    # Custom constructor that checks for duplicate keys
+    # (from https://gist.github.com/pypt/94d747fe5180851196eb)
+    def no_duplicates_constructor(
+        loader: yaml.Loader, node: yaml.Node, deep: bool = False
+    ) -> Any:
+        mapping: Dict[str, Any] = {}
+        for key_node, value_node in node.value:
+            key = loader.construct_object(key_node, deep=deep)
+            value = loader.construct_object(value_node, deep=deep)
+            if key in mapping:
+                raise yaml.constructor.ConstructorError(
+                    "while constructing a mapping",
+                    node.start_mark,
+                    f"found duplicate key {key}",
+                    key_node.start_mark,
+                )
```

Comment:
To avoid breaking compatibility, I think I prefer a warning instead.
What do you think about something like this?
```
warnings.warn(category=UserWarning, message=...)
```
Also, the warning message is very technical, we can make it clearer.
Maybe by inverting the order: "Duplicate key found while constructing a mapping" (along with embedding anything helpful in there).

Reply:
The format here is specific to `ConstructionError`. The final error reported to the user becomes:
```
yaml.constructor.ConstructorError: while constructing a mapping
  in "/var/folders/37/wpf0jmtx63b0spk4ps1pfbhr0000gn/T/tmp6tqvg9hq", line 3, column 3
found duplicate key b
  in "/var/folders/37/wpf0jmtx63b0spk4ps1pfbhr0000gn/T/tmp6tqvg9hq", line 5, column 3
```

Reply:
I'd argue for making this an error as it seems to be a violation of the YAML spec. A warning can be missed, and in particular when used from Hydra (where it's normal to override specific configuration keys from e.g. the command-line) it seems like it's really easy to introduce misconfigs this way. My specific problem was something like this:

```!yaml
foo:
    bar: 1
    baz: 2
    items:
        - bla
        - blubb

# other items...

# intention: change foo.baz only
foo:
    baz: 3
```

With the default behavior `yaml.load()`, you'll end up with `{'foo': {'baz': 3}}`.

Reply:
> 
> 
> The format here is specific to `ConstructionError`. The final error reported to the user becomes:
> 
> ```
> yaml.constructor.ConstructorError: while constructing a mapping
>   in "/var/folders/37/wpf0jmtx63b0spk4ps1pfbhr0000gn/T/tmp6tqvg9hq", line 3, column 3
> found duplicate key b
>   in "/var/folders/37/wpf0jmtx63b0spk4ps1pfbhr0000gn/T/tmp6tqvg9hq", line 5, column 3
> ```

What does it look like if you construct from a string?

Reply:
In that case you'll also get excerpts from the actual string:
```
yaml.constructor.ConstructorError: while constructing a mapping
  in "<unicode string>", line 2, column 1:
    a:
    ^
found duplicate key a
  in "<unicode string>", line 4, column 1:
    a:
    ^
```